### PR TITLE
Faster float conversion operations

### PR DESCRIPTION
**Pull request by m-ou-se**
This supersedes https://github.com/rust-lang/compiler-builtins/pull/370

This includes https://github.com/rust-lang/compiler-builtins/pull/463

### CODE REVIEW

**Comment by m-ou-se** [comment]
Currently, `compiler_builtins` contains two cases where a builtin is implemented by the native conversion on x86_64:

https://github.com/rust-lang/compiler-builtins/blob/4117da3b2a97061730732041ba2ef0c1edbe0e53/src/float/conv.rs#L91-L95

I'm a bit confused by that. If `i as f32` results in native code and not a call to compiler-builtins, then there doesn't seem to be much reason to optimize the version in compiler-builtins.

The version in this PR does not have those special cases. I can put them back if someone thinks they are actually useful.

**Comment by bjorn3** [comment]
> I'm a bit confused by that. If i as f32 results in native code and not a call to compiler-builtins, then there doesn't seem to be much reason to optimize the version in compiler-builtins.

`as f32` results in an intrinsic call on platforms that don't have a native version of this.

> The version in this PR does not have those special cases. I can put them back if someone thinks they are actually useful.

I think it should be put back.

**Comment by m-ou-se** [comment]
> `as f32` results in an intrinsic call on platforms that don't have a native version of this.

Yes.. That's what I said. The special case is for a platform that _does_ have a native version, which means this function wouldn't be called when someone does `as f64`.

> I think it should be put back.

Why? When would it be used on x86_64?

**Comment by bjorn3** [comment]
Looks like it is also used in case of +soft-float. I guess that special case wouldn't be valid then though.

File: src/float/conv.rs [hunk 1]
```diff
@@ -1,286 +1,351 @@
...
+///
+/// The algorithm is explained here: https://blog.m-ou.se/floats/
+mod int_to_float {
+    pub fn u32_to_f32_bits(i: u32) -> u32 {
+        if i == 0 {
+            return 0;
+        }
+        let n = i.leading_zeros();
+        let a = i << n >> 8; // Significant bits, with bit 24 still in tact.
+        let b = i << n << 24; // Insignificant bits, only relevant for rounding.
```
[the remark this answers is not in the document]
Comment (m-ou-se) [reply src/float/conv.rs]: Updated!

**Comment by nickdesaulniers** [comment]
Nice blog post; does it make sense to put any of these implementation in llvm's compiler-rt?

**Comment by m-ou-se** [comment]
> does it make sense to put any of these implementation in llvm's compiler-rt?

@nickdesaulniers Sure! I don't have time for that, but if someone has time to port these implementations, that might benefit a lot of people :)

**Pull request merged**